app/controllers/community_memberships_controller.rb

```diff
@@ -9,102 +9,210 @@ class CommunityMembershipsController < ApplicationController
+    end
+  end
 
-    # If the community doesn't have any members, make the first one an admin
-    if @current_community.members.count == 0
-      @community_membership.admin = true
+  def validate_terms(consent:, community:)
+    if consent == "on"
+      Result::Success.new(community.consent)
+    else
+      Result::Error.new("Consent not accepted", reason: :consent_not_given)
     end
+  end
 
-    # This is reached only if requirements are fulfilled
-    if @community_membership.save
-      session[:fb_join] = nil
-      session[:invitation_code] = nil
-      # If invite was used, reduce usages left
-      invitation.use_once! if invitation.present?
+  def update_membership!(membership:, invitation_code:, email_address:, consent:, user:, community:)
+    make_admin = community.members.count == 0 # First member is the admin
 
-      Delayed::Job.enqueue(CommunityJoinedJob.new(@current_user.id, @current_community.id))
-      Delayed::Job.enqueue(SendWelcomeEmail.new(@current_user.id, @current_community.id), priority: 5)
-      flash[:notice] = t("layouts.notifications.you_are_now_member")
+    begin
+      ActiveRecord::Base.transaction do
+        if email_address.present?
+          Email.create!(person_id: user.id, address: email_address, community_id: community.id)
+        end
+
+        m_invitation = Maybe(invitation_code).map { |code| Invitation.find_by(code: code) }
+        m_invitation.each { |invitation|
+          invitation.use_once!
+        }
+
+        attrs = {
+          consent: consent,
+          invitation: m_invitation.or_else(nil),
+          status: "accepted"
+        }
+
+        attrs[:admin] = true if make_admin
+
+        membership.update_attributes!(attrs)
+      end
+
+      Result::Success.new(membership)
+    rescue
+      Result::Error.new("Updating membership failed", reason: :update_failed, errors: membership.errors.full_messages)
+    end
+  end
+
+  def ensure_pending_consent
+    if membership.nil?
+      report_missing_membership(@current_user, @current_community)
+    elsif membership.accepted?
+      flash[:notice] = t("layouts.notifications.you_are_already_member")
+      redirect_to root
+    elsif !membership.pending_consent?
       redirect_to root
-    else
-      flash[:error] = t("layouts.notifications.joining_community_failed")
-      logger.info { "Joining a community failed, because: #{@community_membership.errors.full_messages}" }
-      render :action => :new
     end
   end
 
-  def access_denied
+  def report_missing_membership(user, community)
+    ArgumentError.new("User doesn't have membership. Don't know how to continue. person_id: #{user.id}, community_id: #{community.id}")
+  end
+
+  def membership
+    # TODO This should be changed to @current_user.community_membership when the relation is
+    # changed to has_one instead of has_many
+    @membership ||= @current_user.community_memberships.find_by(community_id: @current_community.id)
```

Comment:
This is already in place in master. `community_membership` can be called directly and the `TODO` can be removed.